pkg/cosign/tuf/client.go

```diff
@@ -516,36 +502,162 @@ func cachedTargetsDir(cacheRoot string) string {
 	return filepath.Join(cacheRoot, "targets")
 }
 
-type targetImpl interface {
-	Get(string) ([]byte, error)
-	setImpl
+// Local store implementations
+func newLocalStore() (client.LocalStore, error) {
+	if noCache() {
+		return client.MemoryLocalStore(), nil
+	}
+	tufDB := filepath.Join(rootCacheDir(), "tuf.db")
+	local, err := tuf_leveldbstore.FileLocalStore(tufDB)
+	if err != nil {
+		return nil, fmt.Errorf("creating cached local store: %w", err)
+	}
+	return local, nil
 }
 
-type setImpl interface {
-	Set(string, []byte) error
+type localInitFunc func() (client.LocalStore, error)
+
+type embeddedLocalStore struct {
+	// This is where updated repository info will be written to, if needed.
+	writeable     client.LocalStore
+	writeableFunc localInitFunc
 }
 
-type memoryCache struct {
-	targets map[string][]byte
+var GetEmbedded = func() fs.FS {
+	return embeddedRootRepo
 }
 
-func (m *memoryCache) Set(p string, b []byte) error {
-	if m.targets == nil {
-		m.targets = map[string][]byte{}
+func wrapEmbeddedLocal(s localInitFunc) client.LocalStore {
+	return &embeddedLocalStore{writeableFunc: s, writeable: nil}
+}
+
+func (e embeddedLocalStore) GetMeta() (map[string]json.RawMessage, error) {
+	if e.writeable != nil {
+		// We are using a writeable store, so use that.
+		return e.writeable.GetMeta()
+	}
+	// We haven't needed to create or write new metadata, so get the embedded metadata.
```

Comment:
Talking to myself for a bit...

I wonder whether it'd be clearer to separate out the embedded `LocalStore` implementation?

That is, you have an `embeddedLocalStore` which *just* contains the code to read the meta from the embedded.FS and `SetMeta()` returns an err always. This can be combined in a `wrapperLocalStore` or something.

This would only really move out the bottom part of GetMeta, so maybe just pull that into a private function `getMetaEmbedded`? I always prefer extracting code into methods vs. comments, because it's self-documenting and comments aren't clear whether they refer to the next line or the next 10 lines, but this is definitely personal preference.

Reply:
I like that idea, done, but it does require making both the `wrappedEmbeddedLocalStore` and the `embeddedLocalStore` into impls of `client.LocalStore`